lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -0,0 +1,151 @@
+from http import HTTPStatus
+from logging import Logger
+
+from flask.json import jsonify
+from flask.wrappers import Request, Response
+from flask_restful import abort
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    ApiBase,
+    API_OPERATION,
+    JSON,
+    Parameter,
+    ParamType,
+    Schema,
+    SchemaError,
+)
+from pbench.server.database.database import Database
+from pbench.server.database.models.datasets import (
+    Dataset,
+    MetadataError,
+)
+
+
+class DatasetsList(ApiBase):
+    """
+    API class to list datasets based on PostgreSQL metadata
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("name", ParamType.STRING),
+                Parameter("owner", ParamType.USER),
+                Parameter("access", ParamType.ACCESS),
+                Parameter("controller", ParamType.STRING),
+                Parameter("start", ParamType.DATE),
+                Parameter("end", ParamType.DATE),
+                Parameter(
+                    "metadata",
+                    ParamType.LIST,
+                    element_type=ParamType.KEYWORD,
+                    keywords=ApiBase.METADATA,
+                    required=False,
+                ),
+            ),
+            role=API_OPERATION.READ,
+        )
+
+    def _get(self, json_data: JSON, request: Request) -> Response:
+        """
+        Get a list of datasets matching a set of criteria.
+
+        NOTE: This does not rely on a JSON payload to identify the dataset and
+        desired metadata keys; instead we rely on URI query parameters.
+
+        TODO: The dispatcher is currently merging JSON request body and Flask
+        URI keywords ... it probably makes sense to also merge request query
+        parameters and then validate them all together; although it might make
+        more sense to validate each against separate schemas.
+
+        GET /api/v1/datasets/list?start=1970-01-01&end=2040-12-31&owner=fred&metadata=dashboard.seen&metadata=server.deletion
+        """
+
+        # We missed automatic schema validation due to the lack of a JSON body;
+        # construct an equivalent JSON body now so we can run it through the
+        # validator.
+        json = {}
+        for key in request.args.keys():
+            if key in self.schema:
+                if self.schema[key].type == ParamType.LIST:
+                    json[key] = request.args.getlist(key)
+                else:
+                    json[key] = request.args.get(key)
```

Comment:
I really hate the way compound comprehensions get blackened. It tends to drive me away from even examining options like this. I'm not really sure I'm that thrilled about breaking out the local helper method, either; and, as you point out, without it black will likely get even nastier.

In theory, I do like using comprehensions ... at least when they're compact and not easily uglified.

Reply:
In my case, I have a strong affinity for the ternary operator (for a number of reasons); however, the Python syntax for it is so clunky, and the fact that Black almost inevitably formats it "wrong", has made me start to prefer using an `if` _statement_ instead.  (Why couldn't Python have used `?:` like other languages?...or, failing that, why does it _require_ the `else`?...Perl allows conditional assignments without it...not that we love Perl, but....)  However, the "ternary" expression fits on one line in the helper function, and it's encapsulated so that I don't have to look at it if I don't want to, so I think the result is pretty good.  Nevertheless, I do share some of your disregard for the complexity of the comprehension, so, it's no big deal.  (I'm spurred toward considering them because I dislike the idiom of initializing the list or dict to empty brackets and then repeatedly adding stuff to it when I could just "compose" the list in place....)